### PR TITLE
Validation: Port Reference Implementation

### PR DESCRIPTION
**Pull request by ianhoffman**
Hi! This PR begins to port [the reference implementation](https://github.com/graphql/graphql-js/tree/main/src/validation) of validation. In short, the reference implementation works as follows:

* The programmer defines _rules_ which run on certain AST nodes like Fields or Inline Fragments.
* The framework traverses the AST, running the correct rules for each node.
* While traversing the AST, the framework simultaneously adjusts its position in the schema. Thus, when encountering a Field AST node, the framework will point to the corresponding schema field (should such a field exist).
* Each rule reports any errors to the validator.

There are some 25 rules in the reference implementation, but we may not need to port them all because Hack itself might ensure that some constraints cannot be violated. For this PR, I aimed to port two rules as a proof of concept:

* `ScalarLeafsRule` ([source](https://github.com/graphql/graphql-js/blob/9b0514a093135e9437acbb858f642c49b290728e/src/validation/rules/ScalarLeafsRule.js)): Each non-leaf field must include a selection set, and each leaf field must not include a selection set.
* `FieldsOnCorrectTypeRule` ([source](https://github.com/graphql/graphql-js/blob/9b0514a093135e9437acbb858f642c49b290728e/src/validation/rules/FieldsOnCorrectTypeRule.js)): You may not select fields which do not exist on the parent type.

These rules both work! However, a _lot_ more remains to be done:

* Many rules depend on traversing the AST to recover all variables present in a request. We'll need to implement this traversal.
* We don't report the line and position at which an error occurred. We'll want to store the location of each node when parsing. It's a bummer that Graphpinator doesn't store this info today.
* The reference implementation allows visitors to edit the AST. Needless to say, I'm not sure how we can support this in Hack. I'm also not sure whether the reference implementation of validation uses this editing functionality. 
* And of course there are ~23 more rules to port.

HOWEVER, I want to share this because I'm looking for feedback:

* Is this a good direction in which to go? I can also see us straight-up porting the Graphpinator [Normalizer](https://github.com/infinityloop-dev/graphpinator/tree/master/src/Normalizer) implementation. That implementation is a bit less flexible than the reference implementation — for example, it won't be as easy to extend with user-defined rules — but it might lend itself better to our codebase, which is already based on Graphpinator.
  * If we do go in this direction instead of porting the Graphpinator Normalizer, will we need separate functionality to normalize each request? If so, is this implementation worth it?
* How can I make the visitor implementations less awful? 
  * Right now we need to manually add a `visit` method for each Graphpinator Parsed Node type, and the type system isn't enforcing that we don't skip a node. I think an `accept` method on each node might be a more type-safe way of doing this — thoughts?
  * Similarly, for the `ParallelVisitor`, there's no way to make sure that I'm correctly overriding every method in the base class. (In fact, I'm pretty sure I missed some.) I don't see how the type checker can help here short of adding a *lot* more boilerplate.
  * In general, is there a more Hack-friendly way to write these visitors? Leveraging generics, perhaps?

I know this is a long one so please take your time! 🙇‍♂️

### CODE REVIEW

File: src/Codegen/Builders/ObjectBuilder.hack [hunk 1]
```diff
@@ -58,7 +58,7 @@ final class ObjectBuilder<TField as IFieldBuilder> extends OutputTypeBuilder<\Sl
                 },
             )
             ->addDefault()
-            ->addLine("throw new \Exception('Unknown field: '.\$field_name);")
+            ->addLine("return null;")
```
Comment (ianhoffman) [inline src/Codegen/Builders/ObjectBuilder.hack:61]: Moving this into the implementation of `resolveAsync` so that we can use `getFieldDefinition` to determine whether a field is valid on a parent.

File: src/Codegen/Builders/ObjectBuilder.hack [hunk 2]
```diff
@@ -46,7 +46,7 @@ final class ObjectBuilder<TField as IFieldBuilder> extends OutputTypeBuilder<\Sl
     private function generateGetFieldDefinition(HackCodegenFactory $cg): CodegenMethod {
         $method = $cg->codegenMethod('getFieldDefinition')
             ->setPublic()
-            ->setReturnType('GraphQL\\IFieldDefinition<this::THackType>')
+            ->setReturnType('?GraphQL\\IResolvableFieldDefinition<this::THackType>')
```
Comment (ianhoffman) [inline src/Codegen/Builders/ObjectBuilder.hack:49]: Introduced `IFieldDefinition` which is a non-generic interface extended by `IResolvableFieldDefinition`. I think we won't need this once introspection is live.

File: src/BaseSchema.hack [hunk 3]
```diff
@@ -21,4 +21,20 @@ abstract class BaseSchema {
     ): Awaitable<ValidFieldResult<?dict<string, mixed>>> {
         return new ValidFieldResult(null);
     }
+
+    <<__MemoizeLSB>>
+    public static function getQueryType(): Types\ObjectType {
```
Comment (ianhoffman) [inline src/BaseSchema.hack:26]: Obviously this is lame. I just wanted to minimize the diff. I think @mwildehahn 's introspection PR already fixes these TODOs.

File: src/FieldDefinition.hack [hunk 4]
```diff
@@ -1,14 +1,18 @@
 namespace Slack\GraphQL;
 
-interface IFieldDefinition<TParent> {
+interface IFieldDefinition {
+    public function getType(): Types\IOutputType;
```
Comment (ianhoffman) [inline src/FieldDefinition.hack:4]: Once introspection is live, I can refactor this - maybe it should return `__Type`?

**Comment by ianhoffman** [comment]
> How can I make the visitor implementations less awful?

Thought of an alternative which may be a bit better.

Instead of having one enter and leave method per node type, we can add the following to `ASTVisitor`:

```hack

    private dict<string, IVisitorCallback> $enterCallbacks = dict[];
    private dict<string, IVisitorCallback> $leaveCallbacks = dict[];

    public function onEnter<<<__Enforceable>> reify T as nonnull>((function (T): void) $callback): this {
        $node = \HH\ReifiedGenerics\get_classname<T>();
        $this->enterCallbacks[$node] = new VisitorCallback<T>($callback);
        return $this;
    }

    public function onLeave<<<__Enforceable>> reify T as nonnull>((function (T): void) $callback): this {
        $node = \HH\ReifiedGenerics\get_classname<T>();
        $this->leaveCallbacks[$node] = new VisitorCallback<T>($callback);
        return $this;
    }


    public function enter<<<__Enforceable>> reify T as nonnull>(T $node): void {
        $this->invoke<T>($node, $this->enterCallbacks);
    }

    public function leave<<<__Enforceable>> reify T as nonnull>(T $node): void {
        $this->invoke<T>($node, $this->leaveCallbacks);
    }

    private function invoke<<<__Enforceable>> reify T as nonnull>(
        T $node,
        dict<string, IVisitorCallback> $callbacks
    ): void {
        $classname = \HH\ReifiedGenerics\get_classname<T>();
        $callback = $callbacks[$classname] ?? null;
        if ($callback) {
            $callback as VisitorCallback<T>;
            $callback->invoke($node);
        }
    }
```

`IVisitorCallback` and `VisitorCallback` are as follows:

```hack
interface IVisitorCallback {}

final class VisitorCallback<<<__Enforceable>> reify T as nonnull> implements IVisitorCallback {
    public function __construct(private (function (T): void) $callback) {}

    public function invoke(T $node): void {
        $cb = $this->callback;
        $cb($node);
    }
}
```

Then the visit methods just invoke `enter<T>` and `leave<T>`:

```hack
    private function visitField(Parser\Field\Field $node): void {
        $this->enter<Parser\Field\Field>($node);
        $fieldset = $node->getFields();
        if ($fieldset) {
            $this->visitFieldSet($fieldset);
        }
        foreach ($node->getArguments() ?? dict[] as $argument) {
            $this->visitArgumentValue($argument);
        }
        foreach ($node->getDirectives() ?? vec[] as $directive) {
            // TODO: Directives
        }
        $this->leave<Parser\Field\Field>($node);
    }
```

Rules register their callbacks when they init:

```hack
final class FieldsOnCorrectTypeRule extends ValidationRule {
    protected function registerCallbacks(): void {
        $this->onEnter<\Graphpinator\Parser\Field\Field>($field ==> {
            $type = $this->context->getParentType();
            if ($type) {
                $field_def = $this->context->getFieldDef();
                if (!$field_def) {
                    // TODO: Suggestions
                    $this->context->addError(new \Slack\GraphQL\UserFacingError(
                        'Cannot query field "%s" on type "%s".',
                        $field->getName(),
                        $type->getName(),
                    ));
                }
            }
        });
    }
}
```

If this is all a bit hard to follow, I can push a separate PR with this approach (or refactor later). WDYT?

File: src/Codegen/Builders/ObjectBuilder.hack (hunk 2 above)
Comment (jjergus) [reply src/Codegen/Builders/ObjectBuilder.hack:49]: We'll need a resolver-less field definition class sooner or later anyway (e.g. interface field definitions should not have resolvers) so this is probably good.

**Comment by jjergus** [comment]
> We don't report the line and position at which an error occurred. We'll want to store the location of each node when parsing. It's a bummer that Graphpinator doesn't store this info today.

This would be nice but not strictly required (the 'location' field for errors is optional). In many cases we can report the 'path' instead, like we do for execution errors currently.

> The reference implementation allows visitors to edit the AST. Needless to say, I'm not sure how we can support this in Hack. I'm also not sure whether the reference implementation of validation uses this editing functionality.

I don't think it's used for validation. If we later need that for something else, we'd probably want the editing-visitor to return a new AST instead of modifying it in-place.

File: src/__Private/Visitor/ASTVisitor.hack [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        foreach ($node->getDirectives() ?? vec[] as $directive) {
+            // TODO: Directives
+        }
+        foreach ($node->getVariables() ?? vec[] as $variable) {
+            $this->visitVariable($variable);
+        }
+        $this->leaveOperation($node);
+    }
+
+    private function visitTypeRef(Parser\TypeRef\TypeRef $node): void {
```
Comment (jjergus) [inline src/__Private/Visitor/ASTVisitor.hack:77]: Some of these feel unnecessarily granular to me. Do we need to visit literally every single AST node?

In this particular example, I don't think any validation rule can do anything useful with a TypeRef without any context (e.g. you can't even validate whether a type name is valid without knowing if we're expecting an input type or output type). So a validation rule like "variable types are valid" would likely override `visitVariableDef` and then call `$node->getType()`, instead of overriding `visitTypeRef` directly; or it could even just override `visitOperation` and call `$operation_node->getVariableDefs()`...

**Review by jjergus — COMMENTED** [review]
I think I like your suggestion to use a single visit/leave method. I'm not sure if we even need any generics, it could also just be something like this?

```Hack
<<__Override>>
public function enter(Node $node): void {
  if ($node is FieldSet) {
    // ...
  } else if ($node is Operation) {
    // ...
  }
}
```

File: src/Types/BaseType.hack [hunk 6]
```diff
@@ -8,6 +8,8 @@ abstract class BaseType {
     // <<GraphQL\Field('name', '')>>
     abstract public function getName(): ?string;
 
+    abstract public function getNamedType(): BaseType;
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Types/BaseType.hack:11]: I think we can have this on just InputType and OutputType (or `IInputType` and `IOutputType`). Happy to rename it too — `unwrapType()` is nice. I'll give that a shot.

File: src/Types/Output/ObjectType.hack [hunk 7]
```diff
@@ -25,8 +27,13 @@ abstract class ObjectType extends NamedOutputType {
 
         $results = await Dict\map_async(
             $child_fields,
-            async $child_field ==>
-                await $this->getFieldDefinition($child_field->getName())->resolveAsync($value, $child_field, $vars),
+            async $child_field ==> {
+                $field_definition = $this->getFieldDefinition($child_field->getName());
+                if ($field_definition is null) {
+                    throw new \Exception('Unknown field: '.$child_field->getName());
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Types/Output/ObjectType.hack:33]: Good call, will update.

File: src/Validation/ValidationContext.hack [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    private vec<\Slack\GraphQL\UserFacingError> $errors = vec[];
+
+    public function __construct(
+        private classname<\Slack\GraphQL\BaseSchema> $schema,
+        // TODO: Pass in parsed AST as well as rules may need it.
+        private TypeInfo $type_info,
+    ) {}
+
+    // TODO: We need to also record the location at which the error occured, but 
+    // Graphpinator does not store this info on the AST. We should change that.
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Validation/ValidationContext.hack:16]: Yeah, good idea. That's what the reference implementation does IIRC.

File: src/Validation/ValidationContext.hack [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    public function addError(\Slack\GraphQL\UserFacingError $error): void {
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Validation/ValidationContext.hack:17]: I like both those suggestions, will adopt!

File: src/Validation/ValidationContext.hack [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        $this->errors[] = $error;
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Validation/ValidationContext.hack:18]: Yup, I'll need to start tracking the path when entering and leaving nodes in `TypeInfo`, but that should be doable as a follow-up.

File: src/__Private/Visitor/ASTVisitor.hack (hunk 5 above)
Comment (ianhoffman) [reply src/__Private/Visitor/ASTVisitor.hack:77]: Yeah good point. Every node is visitable in the reference implementation (I'm pretty sure), so I just mindlessly followed that 😉 . But yeah, I agree that we probably don't need all of these. I'm not sure exactly what we'll need, but I imagine at least the following:

* Operations
* Variables
* Fields
* Fragments
* Directives
* Values (I think?)

But maybe we don't need separate methods for every single type of value, etc.

File: src/__Private/Visitor/ASTVisitor.hack [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+     */
+    final public function runAfter(ASTVisitor $visitor): ASTVisitor {
+        return new ParallelVisitor(vec[$visitor, $this]);
+    }
+
+    // Actual visitor implemention.
+    //
+    // TODO: Is there a better way to enforce that each node be visitable?
+    // Maybe we should force all Graphphinator Parsed AST nodes to implement
+    // an `accept` method and move this logic there?
```
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/__Private/Visitor/ASTVisitor.hack:25]: True. I guess we'll just need to be careful 😓

[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/__Private/Visitor/ASTVisitor.hack:25]: (And have tests.)

**Comment by ianhoffman** [comment]
> I think I like your suggestion to use a single visit/leave method. I'm not sure if we even need any generics, it could also just be something like this?
> 
> ```hack
> <<__Override>>
> public function enter(Node $node): void {
>   if ($node is FieldSet) {
>     // ...
>   } else if ($node is Operation) {
>     // ...
>   }
> }
> ```

Totally, that's simpler and would also work (there's unfortunately no base node class/interface at the moment, but we should create one).

File: src/Validation/ValidationContext.hack (hunk 10 above)
[the remark this answers is not in the document]
Comment (jjergus) [reply src/Validation/ValidationContext.hack:18]: You have the field definition stack, that can probably be used to reconstruct the path

File: src/__Private/Visitor/ASTVisitor.hack (hunk 5 above)
Comment (jjergus) [reply src/__Private/Visitor/ASTVisitor.hack:77]: > Every node is visitable in the reference implementation

AFAIK the reference implementation is meant to be more generic, not just support validation -- so it probably allows a lot of cases we don't really need, at least for now.

File: src/Validation/ValidationContext.hack (hunk 10 above)
[the remark this answers is not in the document]
Comment (ianhoffman) [reply src/Validation/ValidationContext.hack:18]: Good point!

**Review by jjergus — APPROVED** [review]

**Review by mwildehahn — APPROVED** [review]
So cool!

**Review by mwildehahn — APPROVED** [review]
